sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,867 @@
+                         phase;
+            }
+            else {
+                phase = phase - gnrc_netdev->lwmac.last_wakeup;
+            }
+            /* If the relative phase is beyond 4/5 cycle time, go to sleep. */
+            if (phase > (4*RTT_US_TO_TICKS(LWMAC_WAKEUP_INTERVAL_US)/5)) {
+                gnrc_netdev_lwmac_set_quit_rx(gnrc_netdev, true);
+            }
+
+            if (gnrc_netdev_lwmac_get_quit_rx(gnrc_netdev)) {
+                lwmac_set_state(gnrc_netdev, LWMAC_SLEEPING);
+            }
+            else {
+                /* Go back to LISTENING after successful reception */
+                lwmac_set_state(gnrc_netdev, LWMAC_LISTENING);
+            }
+            break;
+        }
+        default:
+            lwmac_rx_update(gnrc_netdev);
+    }
+
+    /* If state has changed, reschedule main state machine */
+    if (state_rx != gnrc_netdev->rx.state) {
+        lwmac_schedule_update(gnrc_netdev);
+    }
+}
+
+static void _tx_management(gnrc_netdev_t *gnrc_netdev)
+{
+    char *tx_success = "";
+    lwmac_tx_state_t state_tx = gnrc_netdev->tx.state;
+
+    switch (state_tx) {
+        case TX_STATE_STOPPED: {
+            gnrc_pktsnip_t *pkt;
+
+            /* If there is packet remaining for retransmission,
+             * retransmit it (i.e., the retransmission scheme of LWMAC). */
+            if (gnrc_netdev->tx.packet != NULL) {
+                LOG(LOG_WARNING, "WARNING: [lwmac] TX %d times retry\n",
+                                 gnrc_netdev->tx.tx_retry_count);
+                gnrc_netdev->tx.state = TX_STATE_INIT;
+                gnrc_netdev->tx.wr_sent = 0;
+                lwmac_tx_update(gnrc_netdev);
+            }
+            else {
+                if ((pkt = gnrc_priority_pktqueue_pop(
+                         &gnrc_netdev->tx.current_neighbor->queue))) {
+                    gnrc_netdev->tx.tx_retry_count = 0;
+                    lwmac_tx_start(gnrc_netdev, pkt, gnrc_netdev->tx.current_neighbor);
+                    lwmac_tx_update(gnrc_netdev);
+                }
+                else {
+                    /* Shouldn't happen, but never observed this case */
+                    lwmac_set_state(gnrc_netdev, LWMAC_SLEEPING);
+                }
+            }
+            break;
+        }
+        case TX_STATE_FAILED: {
+            /* If transmission failure, do not try burst transmissions and quit other
+             * transmission attempts in this cycle for collision avoidance */
+            gnrc_netdev_lwmac_set_tx_continue(gnrc_netdev, false);
+            gnrc_netdev_lwmac_set_quit_tx(gnrc_netdev, true);
+            tx_success = "NOT ";
+            /* Intended fall-through, TX packet will therefore be dropped. No
+             * automatic resending here, we did our best.
+             */
+        }
+        case TX_STATE_SUCCESSFUL: {
+            if (gnrc_netdev->tx.current_neighbor == &(gnrc_netdev->tx.neighbors[0])) {
```

Comment:
move to new function `static void _tx_management_successful(gnrc_netdev_t *gnrc_netdev);`?